sklearn_numba_dpex/common/kernels.py

```diff
@@ -215,13 +310,122 @@ def partial_sum_reduction(
         n_groups = math.ceil(n_groups / (2 * work_group_size))
         global_size = n_groups * work_group_size
         kernel = partial_sum_reduction[global_size, work_group_size]
-        result = dpt.empty(n_groups, dtype=dtype, device=device)
+        result_shape = n_groups if is_1d else (n_rows, n_groups)
+        result = dpt.empty(result_shape, dtype=dtype, device=device)
         kernels_and_empty_tensors_pairs.append((kernel, result))
 
     def sum_reduction(summands):
+        # TODO: manually dispatch the kernels with a SyclQueue
         for kernel, result in kernels_and_empty_tensors_pairs:
             kernel(summands, result)
             summands = result
```

Comment:
It would be more efficient to do the calls to `dpt.empty(result_shape, dtype=dtype, device=device)` lazily inside this loop, instead of allocating all the results buffers ahead of time?

Or maybe the Python GC would add sequential overhead that would kill the performance?

Reply:
By allocating ahead of time we also keep the buffers allocated and re-use the same buffers across all future calls to `sum_reduction`. The buffers will be only be garbage collected when `sum_reduction` is garbage collected. A given instance of `sum_reduction` in our loops is going to be called once per iteration so I think it's more sensible this way.

Reply:
Alright!

Reply:
Maybe add an inline comment to explain this!